api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,138 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitFilterSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitFilterList contains a list of RateLimitFilter resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimitFilter `json:"items"`
+}
+
+// RateLimitFilterSpec defines the desired state of RateLimitFilter.
+// +union
+type RateLimitFilterSpec struct {
+	// Type decides the scope for the RateLimits.
+	// Valid RateLimitType values are:
+	//
+	// * "Global" - In this mode, the rate limits are applied across all Envoy proxy instances.
+	//
+	// +unionDiscriminator
+	Type RateLimitType `json:"type"`
+	// Global rate limit configuration.
+	// +optional
+	Global *GlobalRateLimit `json:"global"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	GlobalRateLimitType RateLimitType = "Global"
+)
+
+// GlobalRateLimit defines the global rate limit configuration.
+type GlobalRateLimit struct {
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RateLimitMatch `json:"matches,omitempty"`
+	Limit   RateLimitValue   `json:"limit"`
+}
+
+// RateLimitMatch specifies the attributes within the traffic flow that can
+// be matched on.
+type RateLimitMatch struct {
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []HeaderMatch `json:"headers,omitempty"`
+}
+
+// HeaderMatch defines the match attributes within the HTTP Headers of the request.
+type HeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name of the HTTP header.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	Name string `json:"name"`
+
+	// Value within the HTTP header. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered equivalent.
+	// Do not set this field when Type="Distinct", implying matching on any/all unique values within the header.
+	// +optional
+	Value *string `json:"value,omitempty"`
+}
+
+// HeaderMatchType specifies the semantics of how HTTP header values should be
+// compared. Valid HeaderMatchType values are:
+//
+//   - "Exact": Use this type to match the exact value of the Value field against the value of the specified HTTP Header.
```

Comment:
the Exact and RegularExpression types here will practically just generate a static descriptor when sent to the RLS telling it that a match occurred, not exposing the "descriptor" API here and allowing users to configure what the RLS will see when such a match occurs might limit the usefulness of this API

(see https://projectcontour.io/docs/v1.23.2/config/api/#projectcontour.io/v1.RequestHeaderValueMatchDescriptor)

Reply:
the exact and regular expression will not be used in the RLS side, but will be used to build [specific action types](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action-headervaluematch) in the RL filter . From a user perspective they just want to match a header:value tuple pair which this API is exposing

Reply:
@sunjayBhatia please note that Contour is defining its API in terms of descriptors (an indirection) whereas Envoy Gateway is defining its API in terms of existing traffic info such as Headers

Reply:
what personas are configuring the Filter vs. who is configuring the RLS? there has to be some coordination here on descriptors, are we planning that EG will deploy the RLS?

Reply:
yes EG will manage RLS infra as well as config (which is a translation from this API)